File: packages/shared/src/protocol.ts
/**
 * Extension <-> desktop app protocol over ws://127.0.0.1:47831.
 * Every frame is a JSON object with a `type` discriminator. Mirrored in Rust
 * by crates/lare-core/src/protocol.rs — keep both in sync (PROTOCOL_VERSION).
 */
import { z } from "zod";
import { EditEventSchema } from "./edits";
import { DifficultySchema } from "./leetcode";

export const ProblemInfoSchema = z.object({
  slug: z.string(),
  frontendId: z.string().nullable(),
  title: z.string(),
  difficulty: DifficultySchema.nullable(),
  url: z.string(),
  language: z.string().nullable(),
});
export type ProblemInfo = z.infer<typeof ProblemInfoSchema>;

export const SubmissionInfoSchema = z.object({
  leetcodeSubmissionId: z.number().nullable(),
  submittedAt: z.number(), // epoch ms
  lang: z.string().nullable(),
  statusDisplay: z.string().nullable(),
  statusCode: z.number().nullable(),
  accepted: z.boolean(),
  runtimeMs: z.number().nullable(),
  runtimePercentile: z.number().nullable(),
  memoryMb: z.number().nullable(),
  memoryPercentile: z.number().nullable(),
  totalCorrect: z.number().nullable(),
  totalTestcases: z.number().nullable(),
  code: z.string().nullable(),
});
export type SubmissionInfo = z.infer<typeof SubmissionInfoSchema>;

// ----- extension -> app -----------------------------------------------------
export const ExtHelloSchema = z.object({
  type: z.literal("hello"),
  protocol: z.number().int(),
  extVersion: z.string(),
  userId: z.string().nullable(),
});

export const SessionStartSchema = z.object({
  type: z.literal("session.start"),
  sessionId: z.string(),
  kind: z.enum(["practice", "interview"]),
  scope: z.enum(["session", "problem"]),
  startedAt: z.number(),
  problem: ProblemInfoSchema.nullable(),
  /** Interview only: record the webcam as a PiP track. */
  facecam: z.boolean().default(false),
  /** Interview only: capture the microphone (always true for interviews in v1). */
  mic: z.boolean().default(true),
});

export const ProblemOpenSchema = z.object({
  type: z.literal("problem.open"),
  sessionId: z.string(),
  sessionProblemId: z.string(),
  at: z.number(),
  problem: ProblemInfoSchema,
});

export const EditsBatchSchema = z.object({
  type: z.literal("edits.batch"),
  sessionId: z.string(),
  sessionProblemId: z.string(),
  slug: z.string(),
  events: z.array(EditEventSchema),
});

export const SubmissionMsgSchema = z.object({
  type: z.literal("submission"),
  sessionId: z.string(),
  sessionProblemId: z.string(),
  submission: SubmissionInfoSchema,
});

export const ExtToAppSchema = z.discriminatedUnion("type", [
  ExtHelloSchema,
  SessionStartSchema,
  z.object({ type: z.literal("session.pause"), sessionId: z.string(), at: z.number() }),
  z.object({ type: z.literal("session.resume"), sessionId: z.string(), at: z.number() }),
  z.object({ type: z.literal("session.end"), sessionId: z.string(), at: z.number() }),
  ProblemOpenSchema,
  EditsBatchSchema,
  SubmissionMsgSchema,
  z.object({ type: z.literal("ping"), at: z.number() }),
]);
export type ExtToApp = z.infer<typeof ExtToAppSchema>;

// ----- app -> extension -----------------------------------------------------
export const RecordingStateSchema = z.enum([
  "idle",
  "starting",
  "recording",
  "paused",
  "stopping",
  "error",
]);
export type RecordingState = z.infer<typeof RecordingStateSchema>;

export const AppToExtSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("hello.ack"),
    protocol: z.number().int(),
    appVersion: z.string(),
    userId: z.string().nullable(),
    /** True when screen recording permission is granted and devices are available. */
    recordingCapable: z.boolean(),
  }),
  z.object({
    type: z.literal("recording.state"),
    sessionId: z.string().nullable(),
    state: RecordingStateSchema,
    /** Epoch ms when media time 0 occurred; null until recording. */
    startedAt: z.number().nullable(),
    message: z.string().nullable().optional(),
  }),
  z.object({
    type: z.literal("error"),
    code: z.enum([
      "not_signed_in",
      "user_mismatch",
      "permission_denied",
      "already_recording",
      "recording_failed",
      "bad_message",
      "unsupported_protocol",
    ]),
    message: z.string(),
  }),
  z.object({ type: z.literal("pong"), at: z.number() }),
]);
export type AppToExt = z.infer<typeof AppToExtSchema>;

export function encode(msg: ExtToApp | AppToExt): string {
  return JSON.stringify(msg);
}

export function decodeExtToApp(raw: string): ExtToApp | null {
  try {
    const r = ExtToAppSchema.safeParse(JSON.parse(raw));
    return r.success ? r.data : null;
  } catch {
    return null;
  }
}

export function decodeAppToExt(raw: string): AppToExt | null {
  try {
    const r = AppToExtSchema.safeParse(JSON.parse(raw));
    return r.success ? r.data : null;
  } catch {
    return null;
  }
}

File: crates/lare-core/src/protocol.rs
//! Extension <-> desktop WebSocket protocol (mirrors packages/shared/src/protocol.ts).

use serde::{Deserialize, Serialize};

use crate::edits::EditEvent;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Difficulty {
    Easy,
    Medium,
    Hard,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProblemInfo {
    pub slug: String,
    pub frontend_id: Option<String>,
    pub title: String,
    pub difficulty: Option<Difficulty>,
    pub url: String,
    pub language: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubmissionInfo {
    pub leetcode_submission_id: Option<u64>,
    pub submitted_at: u64,
    pub lang: Option<String>,
    pub status_display: Option<String>,
    pub status_code: Option<i32>,
    pub accepted: bool,
    pub runtime_ms: Option<f64>,
    pub runtime_percentile: Option<f64>,
    pub memory_mb: Option<f64>,
    pub memory_percentile: Option<f64>,
    pub total_correct: Option<i64>,
    pub total_testcases: Option<i64>,
    pub code: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SessionKind {
    Practice,
    Interview,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SessionScope {
    Session,
    Problem,
}

fn default_true() -> bool {
    true
}

/// Messages sent by the extension to the desktop app.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ExtToApp {
    #[serde(rename = "hello")]
    Hello {
        protocol: u32,
        #[serde(rename = "extVersion")]
        ext_version: String,
        #[serde(rename = "userId")]
        user_id: Option<String>,
    },
    #[serde(rename = "session.start", rename_all = "camelCase")]
    SessionStart {
        session_id: String,
        kind: SessionKind,
        scope: SessionScope,
        started_at: u64,
        problem: Option<ProblemInfo>,
        #[serde(default)]
        facecam: bool,
        #[serde(default = "default_true")]
        mic: bool,
    },
    #[serde(rename = "session.pause", rename_all = "camelCase")]
    SessionPause { session_id: String, at: u64 },
    #[serde(rename = "session.resume", rename_all = "camelCase")]
    SessionResume { session_id: String, at: u64 },
    #[serde(rename = "session.end", rename_all = "camelCase")]
    SessionEnd { session_id: String, at: u64 },
    #[serde(rename = "problem.open", rename_all = "camelCase")]
    ProblemOpen {
        session_id: String,
        session_problem_id: String,
        at: u64,
        problem: ProblemInfo,
    },
    #[serde(rename = "edits.batch", rename_all = "camelCase")]
    EditsBatch {
        session_id: String,
        session_problem_id: String,
        slug: String,
        events: Vec<EditEvent>,
    },
    #[serde(rename = "submission", rename_all = "camelCase")]
    Submission {
        session_id: String,
        session_problem_id: String,
        submission: SubmissionInfo,
    },
    #[serde(rename = "ping")]
    Ping { at: u64 },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RecordingState {
    Idle,
    Starting,
    Recording,
    Paused,
    Stopping,
    Error,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ErrorCode {
    NotSignedIn,
    UserMismatch,
    PermissionDenied,
    AlreadyRecording,
    RecordingFailed,
    BadMessage,
    UnsupportedProtocol,
}

/// Messages sent by the desktop app to the extension.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum AppToExt {
    #[serde(rename = "hello.ack", rename_all = "camelCase")]
    HelloAck {
        protocol: u32,
        app_version: String,
        user_id: Option<String>,
        recording_capable: bool,
    },
    #[serde(rename = "recording.state", rename_all = "camelCase")]
    RecordingState {
        session_id: Option<String>,
        state: RecordingState,
        started_at: Option<u64>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        message: Option<String>,
    },
    #[serde(rename = "error")]
    Error { code: ErrorCode, message: String },
    #[serde(rename = "pong")]
    Pong { at: u64 },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn session_start_wire_format_matches_ts() {
        let raw = r#"{"type":"session.start","sessionId":"s1","kind":"interview","scope":"problem","startedAt":1700000000000,"problem":{"slug":"two-sum","frontendId":"1","title":"Two Sum","difficulty":"Easy","url":"https://leetcode.com/problems/two-sum/","language":"python"},"facecam":true,"mic":true}"#;
        let msg: ExtToApp = serde_json::from_str(raw).unwrap();
        match &msg {
            ExtToApp::SessionStart { session_id, kind, facecam, problem, .. } => {
                assert_eq!(session_id, "s1");
                assert_eq!(*kind, SessionKind::Interview);
                assert!(*facecam);
                assert_eq!(problem.as_ref().unwrap().difficulty, Some(Difficulty::Easy));
            }
            other => panic!("unexpected {other:?}"),
        }
        let back = serde_json::to_string(&msg).unwrap();
        assert!(back.contains(r#""type":"session.start""#));
        assert!(back.contains(r#""startedAt":1700000000000"#));
    }

    #[test]
    fn app_messages_serialize_with_camel_case() {
        let m = AppToExt::HelloAck {
            protocol: 1,
            app_version: "0.1.0".into(),
            user_id: None,
            recording_capable: true,
        };
        let s = serde_json::to_string(&m).unwrap();
        assert_eq!(
            s,
            r#"{"type":"hello.ack","protocol":1,"appVersion":"0.1.0","userId":null,"recordingCapable":true}"#
        );
        let e = AppToExt::Error {
            code: ErrorCode::NotSignedIn,
            message: "sign in".into(),
        };
        assert_eq!(
            serde_json::to_string(&e).unwrap(),
            r#"{"type":"error","code":"not_signed_in","message":"sign in"}"#
        );
    }
}
